xen/arch/riscv/mm.c

```diff
@@ -349,25 +366,173 @@ unsigned long get_upper_mfn_bound(void)
+			unsigned long __load_address = (unsigned long) (load_address); \
+			if (load_addr_start <= __load_address && __load_address < load_addr_end) { \
+				__load_address = __load_address - load_addr_start + linker_addr_start; \
+			} \
+			__load_address; \
+			})
+
+/*
+ * 1) Build the page tables for Xen that map the following:
+ *   1.1)  The physical location of Xen (where the bootloader loaded it)
+ *   1.2)  The link-time location of Xen (where linker expected Xen's
+ *         addresses to be)
+ * 2) Load the page table into the SATP and enable the MMU
+ */
+void __attribute__ ((section(".entry")))
+    _setup_initial_pagetables(unsigned long load_addr_start, unsigned long load_addr_end,
+			 unsigned long linker_addr_start, unsigned long linker_addr_end)
+{
+    unsigned long *second;
+    unsigned long *first;
+    unsigned long *zeroeth;
+
+    clear_pagetables(load_addr_start, linker_addr_start);
+
+    /* Get the addresses where the page tables were loaded */
+    second = (unsigned long *)load_addr(&xen_second_pagetable);
+    first = (unsigned long *)load_addr(&xen_first_pagetable);
+    zeroeth = (unsigned long *)load_addr(&xen_zeroeth_pagetable);
+
+    /* Create a mapping of the load time address range to... the load time address range.
+     * This mapping is used at boot time only.
+     */
+    setup_initial_pagetables(second, first, zeroeth, load_addr_start, load_addr_end, load_addr_start);
+
+    /* Create a mapping of all of Xen's link-time addresses to where they were actually loaded.
+     * This mapping is used at runtime.
+     */
+    setup_initial_pagetables(second, first, zeroeth, 
+                       linker_addr(&_text_start),
+                       linker_addr(&_text_end),
+                       load_addr(&_text_start));
+    setup_initial_pagetables(second, first, zeroeth,
+                       linker_addr(&_init_start),
+                       linker_addr(&_init_end),
+                       load_addr(&_init_start));
+    setup_initial_pagetables(second, first, zeroeth,
+                       linker_addr(&_cpuinit_start),
+                       linker_addr(&_cpuinit_end),
+                       load_addr(&_cpuinit_start));
+    setup_initial_pagetables(second, first, zeroeth,
+                       linker_addr(&_spinlock_start),
+                       linker_addr(&_spinlock_end),
+                       load_addr(&_spinlock_start));
+    setup_initial_pagetables(second, first, zeroeth,
+                       linker_addr(&_rodata_start),
+                       linker_addr(&_rodata_end),
+                       load_addr(&_rodata_start));
+    setup_initial_pagetables(second, first, zeroeth, linker_addr_start, linker_addr_end, load_addr_start);
+
+    /* Ensure page table writes precede loading the SATP */
+    asm volatile ("sfence.vma");
+
+    /* Enable the MMU and load the pagetable */
+    csr_write(satp, (load_addr(xen_second_pagetable) >> PAGE_SHIFT) | SATP_MODE);
```

Comment:
> If the implementation does not provide ASIDs, or software chooses to always use ASID 0,
then after every satp write, software should execute SFENCE.VMA with rs1=x0.

From Section 4.2.1 of the spec. I may be misunderstanding `sfence.vma`, but I think `sfence.vma` should be called after the write to `satp`?

Reply:
I need to dig deeper into `sfence.vma`, the kernel does it both ways: 

`sfence.vma` before writing `satp`: https://elixir.bootlin.com/linux/v5.3.5/source/arch/riscv/kernel/head.S#L131
`sfence.vma` after writing `satp`: https://elixir.bootlin.com/linux/v5.3.5/source/arch/riscv/kernel/head.S#L151

Reply:
It appears that the fence prior or the fence after updating `satp` handle two slightly different cases.  Fencing prior to writing `satp` seems to be used exclusively to ensure that the yet-to-be-activated page table writes have completed before loading the `satp` (and possibly to flush the TLB of stagnant entries), while the fence after writing `satp` is used to ensure that the following explicit/implicit read/writes are all subject to the new page table translations.

I suspect that the reason the kernel doesn't fence after updating the `satp` with the boot page tables is because the boot page tables being switched to include exact mappings of the physical memory that any implicit/explicit reads/writes would affect if they were to bypass the MMU.  With that physical memory mapped in as virtual memory, the effect of those reads/writes is identical (except for memory protections) with or without the `satp` update being yet active.  It does use `sfence.vma` prior to the `satp` update, though, just to resolve any writes to the inactive tables before executing subsequent instructions, and thus avoiding any page faults if those subsequent accesses were to pages whose updates were still in the pipeline.

The context switch for processes doesn't include an `sfence.vma` before the `satp` because the `sfence.vma` following the `satp` update is guaranteed to be mapped in already because it is in the kernel address space which is always mapped in regardless of process context.

Reply:
Great summary! `sfence.vma` makes a bit more sense now. I still haven't figured out ASIDs though :smile:

Reply:
Lol me neither.  I'll add that to the pile of a thousand things that I would like to better understand.